src/orange/g4org/PhysicalVolumeConverter.cc

```diff
@@ -0,0 +1,220 @@
+    PhysicalVolume make_pv(int depth, G4VPhysicalVolume const& pv);
+
+    // Build a daughter
+    void
+    place_daughter(int depth, G4VPhysicalVolume const& g4pv, LogicalVolume* lv);
+
+    // Build all daughters in the queue
+    void build_daughters();
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with options.
+ */
+PhysicalVolumeConverter::PhysicalVolumeConverter(bool verbose)
+    : data_{std::make_unique<Data>()}
+{
+    data_->verbose = verbose;
+}
+
+//---------------------------------------------------------------------------//
+//! Default destructor
+PhysicalVolumeConverter::~PhysicalVolumeConverter() = default;
+
+//---------------------------------------------------------------------------//
+auto PhysicalVolumeConverter::operator()(arg_type g4world) -> result_type
+{
+    CELER_EXPECT(!g4world.GetRotation());
+    CELER_EXPECT(g4world.GetTranslation() == G4ThreeVector(0, 0, 0));
+
+    CELER_LOG(status) << "Converting Geant4 geometry";
+    ScopedProfiling profile_this{"import-geant-geo"};
+    ScopedMem record_mem("PhysicalVolumeConverter.convert");
+    ScopedTimeLog scoped_time;
+
+    // Construct world volume
+    Builder impl{data_.get(), {}};
+    auto world = impl.make_pv(0, g4world);
+    impl.build_daughters();
+    return world;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a physical volume without building daughters.
+ */
+PhysicalVolume
+PhysicalVolumeConverter::Builder::make_pv(int depth,
+                                          G4VPhysicalVolume const& g4pv)
+{
+    PhysicalVolume result;
+
+    result.name = g4pv.GetName();
+    result.copy_number = g4pv.GetCopyNo();
+    result.transform = this->data->make_transform(g4pv.GetFrameTranslation(),
+                                                  g4pv.GetFrameRotation());
+
+    auto* g4lv = g4pv.GetLogicalVolume();
+    if (auto* unrefl_g4lv
+        = G4ReflectionFactory::Instance()->GetConstituentLV(g4lv))
+    {
+        // Replace with constituent volume, and reflect across Z.
+        // See G4ReflectionFactory::CheckScale: the reflection value is
+        // hardcoded to {1, 1, -1}
+        g4lv = unrefl_g4lv;
+        CELER_NOT_IMPLEMENTED("reflecting a placed volume");
+    }
+
+    // Convert logical volume
+    if (CELER_UNLIKELY(data->verbose))
+    {
+        std::clog << std::string(depth, ' ') << "Converting "
+                  << g4lv->GetName() << std::endl;
+    }
+    auto&& [lv, inserted] = this->data->make_lv(*g4lv);
+    if (inserted)
+    {
+        // Queue up daughters for construction
+        auto num_daughters = g4lv->GetNoDaughters();
+        lv->daughters.reserve(num_daughters);
+        for (auto i : range(num_daughters))
+        {
+            G4VPhysicalVolume* g4pv = g4lv->GetDaughter(i);
+            CELER_ASSERT(g4pv);
+            daughter_queue.push_back({depth + 1, g4pv, lv});
```

Comment:
Maybe, but remember this is all for preprocessing, and the Geant4 objects are not very memory efficient anyway. (Also we're just pushing back a pointer rather than dereferencing it and reusing it later). I think having a queue/stack gives us more flexibility for construction since we can switch from breadth-first to depth-first or (as I was originally wondering if we had to do ) reverse depth-first.